wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -0,0 +1,283 @@
+      incomingTxoDAO
+        .findTx(transaction)
+        .flatMap {
+          // no existing elements found
+          case Vector() =>
+            processNewIncomingTx(transaction, confirmations)
+
+          case txos: Vector[IncomingWalletTXO] =>
+            val txoProcessingFutures: Vector[
+              Future[Option[IncomingWalletTXO]]] = txos
+              .map(processExistingIncomingTxo(transaction, confirmations, _))
+
+            Future
+              .sequence(txoProcessingFutures)
+              .map(_.flatten)
+
+        }
+
+    val outgoingTxFut: Future[Vector[IncomingWalletTXO]] = {
+      for {
+        outputsBeingSpent <- incomingTxoDAO.findOutputsBeingSpent(transaction)
+        processed <- FutureUtil.sequentially(outputsBeingSpent)(
+          markAsSpentIfUnspent)
+      } yield processed.flatten.toVector
+
+    }
+
+    val aggregateFut =
+      for {
+        incoming <- incomingTxoFut
+        outgoing <- outgoingTxFut
+      } yield {
+        ProcessTxResult(incoming.toList, outgoing.toList)
+      }
+
+    aggregateFut.failed.foreach { err =>
+      val msg = s"Error when processing transaction=${transaction.txIdBE}"
+      logger.error(msg, err)
+    }
+
+    aggregateFut
+  }
+
+  /** If the given UTXO is marked as unspent, updates
+    * its spending status. Otherwise returns `None`.
+    */
+  private val markAsSpentIfUnspent: (
+      (SpendingInfoDb, IncomingWalletTXO)) => Future[
+    Option[IncomingWalletTXO]] = {
+    case (info, out) =>
+      if (out.spent) {
+        Future.successful(None)
+      } else {
+        val updatedF =
+          incomingTxoDAO.update(out.copy(spent = true))
+        updatedF.foreach(
+          updated =>
+            logger.debug(
+              s"Marked utxo=${info.toHumanReadableString} as spent=${updated.spent}")
+        )
+        updatedF.map(Some(_))
+      }
+  }
+
+  /**
+    * Inserts the UTXO at the given index into our DB, swallowing the
+    * error if any (this is because we're operating on data we've
+    * already verified).
+    */
+  private def processUtxo(
+      transaction: Transaction,
+      index: Int): Future[SpendingInfoDb] =
+    addUtxo(transaction, UInt32(index))
+      .flatMap {
+        case AddUtxoSuccess(utxo) => Future.successful(utxo)
+        case err: AddUtxoError =>
+          logger.error(s"Could not add UTXO", err)
+          Future.failed(err)
+      }
+
+  private case class OutputWithIndex(output: TransactionOutput, index: Int)
+
+  /**
+    * @param xs UTXO sequence we want to create
+    *           incoming wallet TXOs for
+    */
+  private def createIncomingTxos(
+      transaction: Transaction,
+      confirmations: Int,
+      xs: Seq[(SpendingInfoDb, OutputWithIndex)]): Vector[IncomingWalletTXO] = {
+    xs.map {
+      case (utxo, OutputWithIndex(out, _)) =>
+        IncomingWalletTXO(
+          confirmations = confirmations,
+          // is this always the case?
+          spent = false,
```

Comment:
Yes, i believe it has to be, at least for this transaction. You can't reference a transaction's outpoint within the transaction it was created in. You end up in a recursive dependency situation.